File: src/main/java/com/yumii/core/Transaction.java
package com.yumii.core;


public record Transaction(com.yumii.core.Transaction.TransactionType type, double amount) {

  public static Transaction deposit(double amount) {
    return new Transaction(TransactionType.DEPOSIT, amount);
  }

  public static Transaction withdraw(double amount) {
    return new Transaction(TransactionType.WITHDRAWAL, amount);
  }

  public enum TransactionType {
    DEPOSIT,
    WITHDRAWAL
  }
}

File: src/core/transaction.rs
use std::io::Error;

#[derive(Clone, Debug)]
pub enum TransactionType {
    DEPOSIT,
    WITHDRAWAL,
}

#[derive(Clone)]
pub struct Transaction {
    pub ttype: TransactionType,
    pub amount: f32,
}

impl Transaction {
    pub fn new(ttype: TransactionType, amount: f32) -> Transaction {
        Self { ttype, amount }
    }

    pub fn deposit(amount: f32) -> Transaction {
        Self::new(TransactionType::DEPOSIT, amount)
    }

    pub fn withdraw(amount: f32) -> Transaction {
        Self::new(TransactionType::WITHDRAWAL, amount)
    }
}

pub trait Transactional {
    fn deposit(&mut self, amount: f32) -> Result<Transaction, Error>;
    fn withdraw(&mut self, amount: f32) -> Result<Transaction, Error>;
}
